File: nim/src/benchmarks/sieve.nim
import std/[math, strutils]
import ../benchmark
import ../helper

type
  Sieve* = ref object of Benchmark
    n: int64
    checksumVal: uint32

proc newSieve(): Benchmark =
  Sieve()

method name(self: Sieve): string = "Etc::Sieve"

method prepare(self: Sieve) =
  self.n = self.config_val("limit")
  self.checksumVal = 0'u32

method run(self: Sieve, iteration_id: int) =
  let lim = self.n.int
  var primes = newSeq[uint8](lim + 1)
  for i in 0..lim:
    primes[i] = 1
  primes[0] = 0
  primes[1] = 0

  let sqrtLimit = int(sqrt(float(lim)))

  for p in 2..sqrtLimit:
    if primes[p] == 1:
      var multiple = p * p
      while multiple <= lim:
        primes[multiple] = 0
        multiple += p

  var lastPrime = 2
  var count = 1

  var n = 3
  while n <= lim:
    if primes[n] == 1:
      lastPrime = n
      count += 1
    n += 2

  self.checksumVal += uint32(lastPrime + count)

method checksum(self: Sieve): uint32 =
  self.checksumVal

registerBenchmark("Etc::Sieve", newSieve)

File: julia/src/benchmarks/Sieve.jl
mutable struct Sieve <: AbstractBenchmark
    n::Int64
    result::UInt32

    function Sieve()
        n = Helper.config_i64("Etc::Sieve", "limit")
        new(n, UInt32(0))
    end
end

name(b::Sieve)::String = "Etc::Sieve"

function run(b::Sieve, iteration_id::Int64)
    lim = Int(b.n)
    primes = fill(UInt8(1), lim + 1)
    primes[1] = UInt8(0)
    primes[2] = UInt8(0)

    sqrt_limit = isqrt(lim)

    for p = 2:sqrt_limit
        if primes[p+1] == 0x01
            start = p * p
            for multiple = start:p:lim
                primes[multiple+1] = 0x00
            end
        end
    end

    last_prime = 2
    count = 1

    for n = 3:2:lim
        if primes[n+1] == 0x01
            last_prime = n
            count += 1
        end
    end

    b.result += UInt32(last_prime + count)
end

checksum(b::Sieve)::UInt32 = b.result
